### PR TITLE
Tag Builder: init

### PR DESCRIPTION
**Pull request by Horki**
@museun hi, this is an initial version of Tag Builder, I created in a special namespace as a helper

TODO:
- should values be unique by key?
- if not, can we have multiple key-value values?
- what sort of sanitization should we have when building tags? (remove empty space, allow ':', ',' in key-value?)

### CODE REVIEW

**Comment by museun** [comment]
* keys should be unique. I imagine 'last' key wins (the normal insertion rules for a hashmap)
    * note, IRCv3 allows for duplicate keys but Twitch does not (yet? they don't send duplicate keys)
* there are some quoting rules about tags, I'll have to find them. the Twitch docs are notably lacking in many areas.

There should be a `merge` function so you can take a pre-parsed message and modify its existed tags with a new set of tags.

This is what I came up with, but there are various edge cases I haven't bothered to handle. I've been exceedingly busy lately:
https://github.com/museun/still_shaken/blob/master/crates/still_shaken/src/bot/test.rs#L264-L318


I think the one provided in this crate should operate directly on `Tags/TagsIndices` though.

File: src/irc/tag_indices.rs [hunk 1]
```diff
@@ -95,3 +103,84 @@ impl IntoOwned<'static> for TagIndices {
...
+    fn test_same_slices() {
+        let tag = TagBuilder::new();
+        let r = tag.with("first", "second").with("first", "second").build();
+        assert_eq!(r, "#first:second,first:second");
+        assert_eq!(tag.count(), 2);
+        assert_ne!(tag.count(), 1);
+        {
+            // TODO: should we be able to append new data after build?
+            let r = tag.with("first", "second").build();
+            assert_eq!(r, "#first:second,first:second,first:second");
```
Comment (museun) [inline src/irc/tag_indices.rs]: The format for tags is `@key1=val1;key2=val2; `

**Comment by museun** [comment]
I ended up rewriting it using some of your design.

I had to do some reworking to support escaping and unescaping. 

I apologize for all of my rewrites of your contributions. This crate is structured weirdly (with a lot of hidden 'magic') that makes it non-trivial (or obvious, to be honest) on how to do things.

Your contributions are motivating me, though. This 'feature' would've stayed unimplemented forever because I would just do it locally for whatever I use this crate for.

**Comment by museun** [comment]
As for your questions: I decided to implement it according to the IRCv3 spec: 
* keys are unique 
    * but it is not an error to replace a key
* values can be empty
* it escapes the keys and values when it encodes the tags
* this won't be exposed off existing types, but through a new 'test' module.
    * the builder would only be used for users to create test messages for their own use.
    * the crate, internally will build up messages outside of tags (via the 'commands' module).

File: src/test/tags_builder.rs [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+        self
+    }
+
+    /// Build the tags reference string and its indices
+    ///
+    /// # Returns
+    /// This returns a tuple that can be used with [`Tags::from_data_indices`](crate::irc::Tags#from_data_indices)
+    ///
+    /// If any empty keys were found, or no keys at all then an error will be returned,
+    pub fn build(self) -> Result<(MaybeOwned<'static>, TagIndices), BuilderError> {
```
Comment (museun) [inline src/test/tags_builder.rs]: From a user perspective this return type is a bit .. lets say weird.

I don't if I should make it more user-friendly -- but Tags are build from a data reference and pre-computed indices.

I could maybe return a simple helper struct which could have a `fn tags(&self) -> Tags<'_>` on it.

Comment (museun) [reply src/test/tags_builder.rs]: I implemented this in 920448e

**Comment by museun** [comment]
I'm satisfied with the design. Merging this

**Comment by Horki** [comment]
@museun that's a lot of work in a short period :+1:
